src/solvers/refinement/string_refinement.cpp

```diff
@@ -1011,31 +1070,75 @@ bool string_refinementt::check_axioms()
     }
   }
 
-  // tells whether one of the not_contains constraint can be violated
-  bool violated_not_contains=false;
+  // Maps from indexes of violated not_contains axiom to a witness of violation
+  std::map<std::size_t, exprt> violated_not_contains;
+
   debug() << "there are " << not_contains_axioms.size()
           << " not_contains axioms" << eom;
-
-  for(auto nc_axiom : not_contains_axioms)
+  for(size_t i=0; i<not_contains_axioms.size(); i++)
   {
-    typet index_type=nc_axiom.s0().length().type();
-    exprt zero=from_integer(0, index_type);
-    exprt witness=generator.get_witness_of(nc_axiom, zero);
+    const string_not_contains_constraintt &nc_axiom=not_contains_axioms[i];
+    const exprt univ_bound_inf=nc_axiom.univ_lower_bound();
+    const exprt univ_bound_sup=nc_axiom.univ_upper_bound();
+    const exprt prem=nc_axiom.premise();
+    const exprt exists_bound_inf=nc_axiom.exists_lower_bound();
+    const exprt exists_bound_sup=nc_axiom.exists_upper_bound();
+    const string_exprt s0=nc_axiom.s0();
+    const string_exprt s1=nc_axiom.s1();
+
+    symbol_exprt univ_var=generator.fresh_univ_index(
+      "not_contains_univ_var", nc_axiom.s0().length().type());
+    exprt witness=generator.get_witness_of(nc_axiom, univ_var);
     exprt val=get(witness);
-    violated_not_contains=true;
+    string_not_contains_constraintt nc_axiom_in_model(
+      get(univ_bound_inf),
+      get(univ_bound_sup),
+      get(prem),
+      get(exists_bound_inf),
+      get(exists_bound_sup),
+      to_string_expr(get(s0)),
+      to_string_expr(get(s1)));
+
+    satcheck_no_simplifiert sat_check;
```

Comment:
So I was thinking how to refactor, and after a conversation with @romainbrenguier, I can't think of a satisfactory method:

1. For the construction of `nc_axiom_in_model`, `get` doesn't properly traverse, and from my investigation this seems like a big enough change to implement it that it should be separate. So the lines creating `nc_axiom_in_model` seem necessary for now.
2. For the initialization and use of the solver, there are only 3 lines associated with creating the solver (and one in `add_negation_of_not_contains_constraint_to_solver`). I don't know if this is definitely the case, but it seems like the minimal set up required. If this were factored out, the 3 line function that may be extended by passing arguments (for example to choose `sat_check` to be `satcheck_no_simplifiert` or `satcheckt`). I feel that then the calls to the function would be just as complicated.
3. For the control flow after running the solver, it is as lightweight as I can imagine. The code in the SAT case is incredibly dependent on the use case. I can imagine a case where we don't want to check any results, or more than one, compared to the one expression there. Additionally, the debug message is highly dependent, and may be desired or not. Finally, the saving of the value in a map seems specific to the `check_axiom` function for now. So if I were to refactor this I can only envisage a function that calls the solver and returns true if SAT and false if UNSAT, which feels like an unnecessary wrapper.

Hence I can't see a way to refactor this that is satisfactory.